contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1047 @@
+      revert InvalidScheduleStartTime({projectId: projectId});
+    }
+    address recipient = scheduleData.supplierAddress;
+    if (recipient == address(0)) {
+      revert RecipientCannotBeZeroAddress();
+    }
+    if (hasRole(SCHEDULE_CREATOR_ROLE, recipient)) {
+      revert RecipientCannotHaveRole({
+        recipient: recipient,
+        role: "SCHEDULE_CREATOR_ROLE"
+      });
+    }
+    if (_scheduleIdToScheduleStruct[projectId].exists) {
+      revert ScheduleExists({scheduleTokenId: projectId});
+    }
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return return value irrelevant
+    _allScheduleIds.add(projectId);
+    RestrictionDuration
+      memory restrictionDuration = getRestrictionDurationForMethodologyAndVersion(
+        scheduleData.methodology,
+        scheduleData.methodologyVersion
+      );
+    if (!restrictionDuration.wasSet) {
+      revert RestrictionDurationNotSet({projectId: projectId});
+    }
+    // slither-disable-next-line unused-return return value irrelevant
+    _addressToScheduleIdSet[recipient].add(projectId);
+    schedule.exists = true;
+    schedule.startTime = scheduleData.startTime;
+    schedule.endTime = scheduleData.startTime + restrictionDuration.duration;
+    schedule.releasedAmountFloor = 0;
+    emit ScheduleCreated(projectId, schedule.startTime, schedule.endTime);
+  }
+
+  /**
+   * @notice Revokes unreleased tokens from a schedule.
+   * This is an *admin* operation callable only by addresses having TOKEN_REVOKER_ROLE
+   * (enforced in `batchRevokeUnreleasedTokenAmounts`)
+   *
+   * @dev Only the maximum revocable number of tokens (unreleased tokens) can be revoked.
+   * Once the tokens have been revoked, the current released amount can never fall below
+   * its current level, even if the linear release schedule of the new amount would cause
+   * the released amount to be lowered at the current timestamp (a floor is established).
+   *
+   * Unlike in the `withdrawFromSchedule` function, here we burn `RestrictedNORI`
+   * from the schedule owner but send that `BridgedPolygonNORI` back to Nori's
+   * treasury or an address of Nori's choosing (the *to* address).
+   * The *claimedAmount* is not changed because this is not a claim operation.
+   */
+  function _revokeUnreleasedTokens(
+    address to,
+    uint256 removalId,
+    uint256 amount
+  ) internal {
+    // slither-disable-next-line calls-loop
+    uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (!schedule.exists) {
+      revert NonexistentSchedule({scheduleId: scheduleId});
+    }
+    uint256 quantityToRevoke;
+    // amount of zero indicates revocation of all remaining tokens.
+    if (amount > 0) {
+      if (!(amount <= revocableQuantityForSchedule(scheduleId))) {
+        revert InsufficientUnreleasedTokens({scheduleId: scheduleId});
+      }
```

Comment:
does this need to happen inside this if block or can it happen higher up since 0 is always <= the minimum `revocableQuantityForSchedule` can be (which is 0)

Reply:
if so, you can assign `quantityToRevoke` in a ternary `quantityToRevoke = amount > 0 ? amount : revocableQuantityForSchedule(scheduleId)`

Reply:
arguably, even that logic should exists in `revocableQuantityForSchedule` and not here

Reply:
that cleaned things up a lot.  the logic around `quantityToRevoke` should stay here though because `revocableQuantityForSchedule` tells you what the total possible revocable quantity for a schedule is and is used in other places, whereas here we just want to know what we're revoking in this case.